lib/clever_tap/entity.rb

```diff
@@ -0,0 +1,83 @@
+class CleverTap
+  class NoDataError < RuntimeError
+    def message
+      'No `data` param provided for Event'
+    end
+  end
+
+  class MissingIdentityError < RuntimeError
+    def message
+      "Couldn'n find `identity` in CleverTap.config or `data`"
+    end
+  end
+
+  class Entity
+    ALLOWED_IDENTITIES = %w(objectId FBID GPID).freeze
+    IDENTITY_STRING = 'identity'.freeze
+    TIMESTAMP_STRING = 'ts'.freeze
+    TYPE_KEY_STRING = 'type'.freeze
+    UPLOAD_LIMIT = 'Needs child class value'.freeze
+    TYPE_VALUE_STRING = 'Needs child class value'.freeze
+
+    class << self
+      def upload_limit
+        self::UPLOAD_LIMIT
+      end
+
+      def all_same_type?(items)
+        items.all? { |i| i.class == self }
+      end
+    end
+
+    def initialize(**args)
+      @data = args[:data]
+      @identity = choose_identity(args)
+      @timestamp = choose_timestamp(args)
+    end
+
+    def to_h
+      put_identity_pair
+        .merge(put_timestamp_pair)
+        .merge(put_type_pair)
+        .merge(put_data)
+    end
+
+    private
+
+    def put_identity_pair
+      raise NoDataError if @data.to_h.empty?
+      raise MissingIdentityError if @identity == '' || @data[@identity].nil?
+      return { @identity => @data[@identity].to_s } if ALLOWED_IDENTITIES.include?(@identity)
+      { IDENTITY_STRING => @data[@identity].to_s }
+    end
+
+    def put_timestamp_pair
+      return {} unless @timestamp
+      { TIMESTAMP_STRING => @timestamp }
+    end
+
+    def put_type_pair
+      { TYPE_KEY_STRING => self.class::TYPE_VALUE_STRING }
+    end
+
+    def put_data
+      raise NoDataError if @data.to_h.empty?
+      @data.delete(@identity) if ALLOWED_IDENTITIES.include? @identity
+      {
+        self.class::DATA_STRING => @data
+      }
+    end
+
+    def choose_identity(args)
+      identity = args[:identity].to_s
+
+      return identity if ALLOWED_IDENTITIES.include?(identity) && @data.to_h.key?(identity)
```

Comment:
`return identity if ALLOWED_IDENTITIES.include?(identity)` is found too much in the code, which means there is some missing concept here

Reply:
maybe move the `ALLOWED_IDENTITIES.include?(identity)` part in a separate method? It's used in a 3 separate cases but I can't think of another good way to improve this, as sometimes it depends on other factors as well. E.g. `&& @data.to_h.key?(identity)` as in this case but not in others.